src/tls/README.md

```diff
@@ -29,40 +31,88 @@ Design goals
-  +---------+   https://machine:9090 |                       |   +------------------------+
-  | Browser |<-----------------------+                       +-->| ws instance for cert A |
-  +---------+     client cert A                                  +------------------------+
++---------+  http://machine:9090                           +------------------------+
+|Browser A|+----------------------+                    +-->|cockpit-ws http instance|
++---------+                       |                    |   +------------------------+
+                                  |                    |
+                                  |                    +   (plain HTTP over Unix socket)
+                                  |                    |
++---------+  https://machine:9090 |    +------------+  |   +------------------------------------+
+|Browser B|+----------------------+--->| cockpit-tls|--+-->|cockpit-ws https instance for cert B|
++---------+    client cert B      |    +------------+  +   +------------------------------------+
+                                  |                    |
+                                  |                    |
++---------+  https://machine:9090 |                    |   +------------------------------------+
+|Browser C|+----------------------+                    +-->|cockpit-ws https instance for cert C|
++---------+    client cert C                               +------------------------------------+
 ```
 
-Current status
---------------
+The startup of these instances and isolating them from one another makes heavy
+use of systemd features.
+
+ * The top-level unit that the admin enables/controls is [cockpit.socket](../ws/cockpit.socket.in),
+   which listens to the port (9090 by default).
+ * That activates [cockpit.service](../ws/cockpit.service.in), which ensures
+   that a server TLS certificate is present and starts `cockpit-tls`, usually
+   as user `cockpit-ws` (for historical reasons, to avoid having to change the
+   ownership of `/etc/cockpit/ws-certs.d/*` on upgrades).
+ * When accepting a connection, cockpit-tls checks if it uses TLS:
+   - If not, it connects to [cockpit-wsinstance-http.socket](../src/ws/cockpit-wsinstance-http.socket.in) or
+     [cockpit-wsinstance-http-redirect.socket](../src/ws/cockpit-wsinstance-http-redirect.socket.in).
+   - If the connection does use TLS, it calculates the fingerprint of the
+     client certificate (using the empty string if there is none), and connects
+     to [cockpit-wsinstance-https-factory.socket](../src/ws/cockpit-wsinstance-https-factory.socket.in).
+     This starts a helper factory process `cockpit-wsinstance-factory` that
+     reads the fingerprint from stdin, and asks systemd to start a new
+     [cockpit-wsinstance-https@fingerprint.socket](../src/ws/cockpit-wsinstance-https@.socket.in)
+     and .service pair.
+ * Each of these instances run in a separate systemd cgroup, as another
```

Comment:
s/run/runs/ now that it's "each of" (which is singular).

I also really liked "its own" as the thing which added the emphasis.  Is there a reason that you didn't go with that?

Reply:
Updated again, thanks!